sgl-pdlb/src/server.rs

```diff
@@ -0,0 +1,341 @@
+                .status(status)
+                .content_type("application/octet-stream")
+                .streaming(body)),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct LBState {
+    strategy_lb: StrategyLB,
+    client: reqwest::Client,
+    log_interval: u64,
+}
+
+impl LBState {
+    pub fn new(lb_config: LBConfig) -> anyhow::Result<Self> {
+        let client = reqwest::Client::builder()
+            .timeout(std::time::Duration::from_secs(lb_config.timeout))
+            .build()?;
+
+        let lb = StrategyLB::new(
+            lb_config.policy,
+            lb_config.prefill_infos,
+            lb_config.decode_infos,
+        );
+        Ok(Self {
+            strategy_lb: lb,
+            client,
+            log_interval: lb_config.log_interval,
+        })
+    }
+
+    async fn route_one(
+        &self,
+        engine_info: &EngineInfo,
+        method: &str,
+        api_path: &str,
+        request: Option<&serde_json::Value>,
+        stream: bool,
+    ) -> Result<ProxyResponse, actix_web::Error> {
+        let url = engine_info.api_path(api_path);
+        let request = request.unwrap_or(&serde_json::Value::Null);
+        let task = match method {
+            "POST" => self.client.post(&url).json(request).send(),
+            "GET" => self.client.get(&url).send(),
+            _ => panic!("Unsupported method: {}", method),
+        };
+        let resp = task
+            .await
+            .map_err(|e| actix_web::error::ErrorBadGateway(e))?;
+        // FIXME: handle error status code (map status code to error)
+        let status = resp.status();
+
+        if stream {
+            let resp_stream = resp.bytes_stream().map(|r| {
+                r.map_err(|e| actix_web::error::ErrorBadGateway(e))
+                    .map(Bytes::from)
+            });
+            Ok(ProxyResponse {
+                status,
+                body: ProxyResponseType::Stream(Box::pin(resp_stream)),
+            })
+        } else {
+            let body = resp
+                .bytes()
+                .await
+                .map_err(|e| actix_web::error::ErrorBadGateway(e))?;
+            Ok(ProxyResponse {
+                status,
+                body: ProxyResponseType::Full(body),
+            })
+        }
+    }
+
+    async fn route_collect(
+        &self,
+        engines: &Vec<EngineInfo>,
+        method: &str,
+        api_path: &str,
+        request: Option<&serde_json::Value>,
+    ) -> Result<Vec<ProxyResponse>, actix_web::Error> {
+        let tasks = engines
+            .iter()
+            .map(|engine| self.route_one(engine, method, api_path, request, false));
+        let responses = join_all(tasks).await;
+        responses
+            .into_iter()
+            .map(|r| r.map_err(|e| actix_web::error::ErrorBadGateway(e)))
+            .collect()
+    }
+
+    async fn generate(
+        &self,
+        api_path: &str,
+        mut req: Box<dyn Bootstrap>,
```

Comment:
```suggestion
        mut req: &impl Bootstrap,
```

(or use full generic)